tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -94,14 +114,257 @@ public int Serialize(ref byte[] bytes, int offset, Span value, IFormatterResolve
                 tagProcessors = tracer.TracerManager?.TagProcessors;
             }
 
-            offset += value.Tags.SerializeTo(ref bytes, offset, value, tagProcessors);
+            offset += SerializeTags(ref bytes, offset, value, value.Tags, tagProcessors);
 
             return offset - originalOffset;
         }
 
+        private int SerializeTags(ref byte[] bytes, int offset, Span span, ITags tags, ITagProcessor[] tagProcessors)
+        {
+            int originalOffset = offset;
+
+            offset += WriteTags(ref bytes, offset, span, tags, tagProcessors);
+            offset += WriteMetrics(ref bytes, offset, span, tags, tagProcessors);
+
+            return offset - originalOffset;
+        }
+
+        // TAGS
+
+        private int WriteTags(ref byte[] bytes, int offset, Span span, ITags tags, ITagProcessor[] tagProcessors)
+        {
+            int originalOffset = offset;
+
+            // Start of "meta" dictionary. Do not add any string tags before this line.
+            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _metaBytes);
+
+            int count = 0;
+
+            // We don't know the final count yet, write a fixed-size header and note the offset
+            var countOffset = offset;
+            offset += MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, offset, 0);
+
+            // add "language=dotnet" tag to all spans, except those that
+            // represents a downstream service or external dependency
+            if (tags is not InstrumentationTags { SpanKind: SpanKinds.Client or SpanKinds.Producer })
+            {
+                count++;
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _languageNameBytes);
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _languageValueBytes);
+            }
+
+            // Write span tags
+            var tagWriter = new TagWriter(this, tagProcessors, bytes, offset);
+            tags.EnumerateTags(ref tagWriter);
+            bytes = tagWriter.Bytes;
+            offset = tagWriter.Offset;
+            count += tagWriter.Count;
+
+            if (span.IsRootSpan)
```

Comment:
No need to do it in this PR, but this needs to change to be "add trace-level tags to the first span in each trace chunk". For now, since most traces are sent in a single chunk, and the root span is usually the first span, it's correct most of the time. But it's wrong when a trace is split into multiple chunks (e.g. partial flushing).

Reply:
I prefer to do all behaviour changes in a different PR. The scope for this PR was only separation of concerns.